index out of range crash

When running `structslop` against `pkg/sql` of [CockroachDB](https://github.com/cockroachdb/cockroach/tree/master/pkg/sql) I got the following:
```
panic: runtime error: index out of range [11] with length 10

goroutine 10501 [running]:
github.com/orijtech/structslop.run.func1(0x138cca0, 0xc006541960)
	/Users/yuzefovich/go/pkg/mod/github.com/orijtech/structslop@v0.0.4/structslop.go:98 +0xbf2
golang.org/x/tools/go/ast/inspector.(*Inspector).Preorder(0xc02cc427a0, 0xc004e81b20, 0x1, 0x1, 0xc004f6fb30)
	/Users/yuzefovich/go/pkg/mod/golang.org/x/tools@v0.0.0-20200917221617-d56e4e40bc9d/go/ast/inspector/inspector.go:77 +0xa2
github.com/orijtech/structslop.run(0xc02cc400c0, 0xc02cc400c0, 0x0, 0x0, 0x0)
	/Users/yuzefovich/go/pkg/mod/github.com/orijtech/structslop@v0.0.4/structslop.go:68 +0x1fa
golang.org/x/tools/go/analysis/internal/checker.(*action).execOnce(0xc02ec706e0)
	/Users/yuzefovich/go/pkg/mod/golang.org/x/tools@v0.0.0-20200917221617-d56e4e40bc9d/go/analysis/internal/checker/checker.go:690 +0x87e
sync.(*Once).doSlow(0xc02ec706e0, 0xc00004f790)
	/usr/local/opt/go/libexec/src/sync/once.go:66 +0xec
sync.(*Once).Do(...)
	/usr/local/opt/go/libexec/src/sync/once.go:57
golang.org/x/tools/go/analysis/internal/checker.(*action).exec(0xc02ec706e0)
	/Users/yuzefovich/go/pkg/mod/golang.org/x/tools@v0.0.0-20200917221617-d56e4e40bc9d/go/analysis/internal/checker/checker.go:579 +0x65
golang.org/x/tools/go/analysis/internal/checker.execAll.func1(0xc02ec706e0)
	/Users/yuzefovich/go/pkg/mod/golang.org/x/tools@v0.0.0-20200917221617-d56e4e40bc9d/go/analysis/internal/checker/checker.go:567 +0x34
created by golang.org/x/tools/go/analysis/internal/checker.execAll
	/Users/yuzefovich/go/pkg/mod/golang.org/x/tools@v0.0.0-20200917221617-d56e4e40bc9d/go/analysis/internal/checker/checker.go:573 +0x125
```

@yuzefovich Thanks for reporting. Would you mind providing details instruction to run? Currently, I got this error when running `structslop ./pkg/sql`:

```
...
/home/cuonglm/sources/cockroach/pkg/sql/exec_factory_util.go:232:10: Factory not declared by package exec
/home/cuonglm/sources/cockroach/pkg/sql/opt_exec_factory.go:49:12: Factory not declared by package exec
/home/cuonglm/sources/cockroach/pkg/sql/plan_opt.go:543:9: Factory not declared by package exec
/home/cuonglm/sources/cockroach/pkg/storage/stacks.go:20:8: could not import C (no metadata for C)
/home/cuonglm/sources/cockroach/pkg/sql/distsql_spec_exec_factory.go:51:12: Factory not declared by package exec
/home/cuonglm/sources/cockroach/pkg/sql/distsql_spec_exec_factory.go:67:83: Factory not declared by package exec
/home/cuonglm/sources/cockroach/pkg/sql/exec_factory_util.go:232:10: Factory not declared by package exec
/home/cuonglm/sources/cockroach/pkg/sql/opt_exec_factory.go:49:12: Factory not declared by package exec
/home/cuonglm/sources/cockroach/pkg/sql/plan_opt.go:543:9: Factory not declared by package exec
/home/cuonglm/sources/cockroach/pkg/server/status/runtime_jemalloc.go:74:8: could not import C (no metadata for C)
/home/cuonglm/sources/cockroach/pkg/sql/show_test.go:889:36: IDENT not declared by package parser
/home/cuonglm/sources/cockroach/pkg/sql/show_test.go:890:21: KeywordsCategories not declared by package lex
structslop: 945 errors during loading
..[cuonglm@cuonglm-ThinkPad-X1-Carbo
```

@yuzefovich it didn't crash for me, it produced this
```shell
$ structslop ./pkg/sql
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/apply_join.go:54:6: struct has size 104 (size class 112), could be 96 (size class 96), rearrange to struct{emptyRight tree.Datums; leftRow tree.Datums; out tree.Datums; rightRows *rowcontainer.RowContainer; curRightRow int; leftRowFoundAMatch bool; done bool} for optimal size (14.29% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/create_view.go:37:21: struct has size 88 (size class 96), could be 80 (size class 80), rearrange to struct{columns colinfo.ResultColumns; viewQuery string; viewName *tree.TableName; planDeps planDependencies; persistence tree.Persistence; dbDesc *dbdesc.Immutable; replace bool; materialized bool; ifNotExists bool} for optimal size (16.67% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:583:18: struct has size 128 (size class 128), could be 104 (size class 112), rearrange to struct{ctx context.Context; spanIter physicalplan.SpanResolverIterator; ExtendedEvalCtx *extendedEvalContext; NodeStatuses map[roachpb.NodeID]NodeStatus; saveFlows func(map[roachpb.NodeID]*execinfrapb.FlowSpec) error; planDepth int; stmtType tree.StatementType; planner *planner; flowID uuid.UUID; ignoreClose bool; isLocal bool; noEvalSubqueries bool; saveDiagramShowInputTypes bool} for optimal size (12.50% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1196:30: struct has size 264 (size class 288), could be 248 (size class 256), rearrange to struct{post execinfrapb.PostProcessSpec; reqOrdering colinfo.ColumnOrdering; spans []roachpb.Span; colsToTableOrdinalMap []int; cols []*descpb.ColumnDescriptor; spec *execinfrapb.TableReaderSpec; estimatedRowCount uint64; desc *tabledesc.Immutable; scanVisibility execinfrapb.ScanVisibility; parallelize bool; reverse bool; containsSystemColumns bool} for optimal size (11.11% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_plan_join.go:28:23: struct has size 360 (size class 384), could be 352 (size class 352), rearrange to struct{post execinfrapb.PostProcessSpec; onExpr execinfrapb.Expression; rightEqCols []uint32; leftMergeOrd execinfrapb.Ordering; rightMergeOrd execinfrapb.Ordering; joinResultTypes []*types.T; joinToStreamColMap []int; leftEqCols []uint32; leftPlan *PhysicalPlan; leftPlanDistribution physicalplan.PlanDistribution; rightPlan *PhysicalPlan; rightPlanDistribution physicalplan.PlanDistribution; joinType descpb.JoinType; rightEqColsAreKey bool; leftEqColsAreKey bool} for optimal size (8.33% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_spec_exec_factory.go:31:29: struct has size 56 (size class 64), could be 48 (size class 48), rearrange to struct{planContexts struct{distPlanCtx *PlanningCtx; localPlanCtx *PlanningCtx}; planner *planner; dsp *DistSQLPlanner; planningMode distSQLPlanningMode; gatewayNodeID roachpb.NodeID; singleTenant bool} for optimal size (25.00% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec_util.go:762:27: struct has size 120 (size class 128), could be 96 (size class 96), rearrange to struct{RunAfterSCJobsCacheLookup func(*jobs.Job); BeforePrepare func(ctx context.Context, stmt string, txn *kv.Txn) error; BeforeExecute func(ctx context.Context, stmt string); AfterExecute func(ctx context.Context, stmt string, err error); AfterExecCmd func(ctx context.Context, cmd Command, buf *StmtBuf); TestingSaveFlows func(stmt string) func(map[roachpb.NodeID]*execinfrapb.FlowSpec) error; BeforeAutoCommit func(ctx context.Context, stmt string) error; StatementFilter StatementFilter; TempObjectsCleanupCh chan time.Time; OnTempObjectsCleanupDone func(); WithStatementTrace func(trace tracing.Recording, stmt string); DisableTempObjectsCleanupOnSessionExit bool; TestingDescriptorValidation bool; DisableAutoCommit bool; DeterministicExplainAnalyze bool} for optimal size (25.00% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/insert.go:49:16: struct has size 528 (size class 576), could be 512 (size class 512), rearrange to struct{ti tableInserter; tabColIdxToRetIdx []int; insertCols []descpb.ColumnDescriptor; resultRowBuffer tree.Datums; rowIdxToTabColIdx []int; checkOrds util.FastIntSet; done bool; rowsNeeded bool; traceKV bool} for optimal size (11.11% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/instrumentation.go:45:28: struct has size 152 (size class 160), could be 128 (size class 128), rearrange to struct{codec keys.SQLCodec; origCtx context.Context; fingerprint string; diagRequestID stmtdiagnostics.RequestID; distribution physicalplan.PlanDistribution; explainPlan *explain.Plan; evalCtx *tree.EvalContext; finishCollectionDiagnostics func(); withStatementTrace func(trace tracing.Recording, stmt string); sp *tracing.Span; explainFlags explain.Flags; outputMode outputMode; discardRows bool; savePlanForStats bool; collectBundle bool; implicitTxn bool; vectorized bool} for optimal size (20.00% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/inverted_join.go:21:23: struct has size 120 (size class 128), could be 112 (size class 112), rearrange to struct{reqOrdering colinfo.ColumnOrdering; columns colinfo.ResultColumns; input planNode; invertedExpr tree.TypedExpr; onExpr tree.TypedExpr; table *scanNode; joinType descpb.JoinType; isFirstJoinInPairedJoiner bool} for optimal size (12.50% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/lookup_join.go:21:21: struct has size 136 (size class 144), could be 120 (size class 128), rearrange to struct{reqOrdering colinfo.ColumnOrdering; eqCols []int; columns colinfo.ResultColumns; input planNode; onCond tree.TypedExpr; table *scanNode; joinType descpb.JoinType; eqColsAreKey bool; isSecondJoinInPairedJoiner bool} for optimal size (11.11% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/scan.go:39:15: struct has size 248 (size class 256), could be 224 (size class 224), rearrange to struct{_ util.NoCopy; colCfg scanColumnsConfig; spans []roachpb.Span; resultColumns colinfo.ResultColumns; cols []*descpb.ColumnDescriptor; reqOrdering colinfo.ColumnOrdering; estimatedRowCount uint64; specifiedIndex *descpb.IndexDescriptor; colIdxMap map[descpb.ColumnID]int; index *descpb.IndexDescriptor; desc *tabledesc.Immutable; hardLimit int64; softLimit int64; lockingWaitPolicy descpb.ScanLockingWaitPolicy; lockingStrength descpb.ScanLockingStrength; noIndexJoin bool; isFull bool; isCheck bool; parallelize bool; disableBatchLimits bool; reverse bool; containsSystemColumns bool} for optimal size (12.50% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/set_zone_config.go:49:24: struct has size 168 (size class 176), could be 160 (size class 160), rearrange to struct{zoneSpecifier tree.ZoneSpecifier; yamlConfig tree.TypedExpr; options map[tree.Name]optionValue; run setZoneConfigRun; allIndexes bool; setDefault bool} for optimal size (9.09% savings)
/Users/emmanuelodeke/go/src/github.com/cockroachdb/cockroach/pkg/sql/update.go:46:16: struct has size 1288 (size class 1408), could be 1280 (size class 1280), rearrange to struct{tu tableUpdater; iVarContainerForComputedCols schemaexpr.RowIndexedVarContainer; rowIdxToRetIdx []int; updateValues tree.Datums; sourceSlots []sourceSlot; computedCols []descpb.ColumnDescriptor; computeExprs []tree.TypedExpr; checkOrds util.FastIntSet; updateColsIdx map[descpb.ColumnID]int; numPassthrough int; traceKV bool; done bool; rowsNeeded bool} for optimal size (9.09% savings)
```

--- a/README.md
+++ b/README.md
@@ -16,7 +16,7 @@ Without Go modules:
 
 ```sh
 $ cd $GOPATH/src/github.com/orijtech/structslop
-$ git checkout v0.0.4
+$ git checkout v0.0.5
 $ go get
 $ install ./cmd/structslop
 ```

--- a/structslop.go
+++ b/structslop.go
@@ -85,22 +85,11 @@ func run(pass *analysis.Pass) (interface{}, error) {
 		}
 
 		var buf bytes.Buffer
-		if err := format.Node(&buf, pass.Fset, atyp); err != nil {
-			return
-		}
-		expr, err := parser.ParseExpr(buf.String())
+		expr, err := parser.ParseExpr(formatStruct(r.optStruct, pass.Pkg.Path()))
 		if err != nil {
 			return
 		}
-		natyp := expr.(*ast.StructType)
-		fl := make([]*ast.Field, styp.NumFields())
-		for i, idx := range r.optIdx {
-			fl[i] = natyp.Fields.List[idx]
-		}
-		natyp.Fields.List = fl
-
-		buf.Reset()
-		if err := format.Node(&buf, token.NewFileSet(), natyp); err != nil {
+		if err := format.Node(&buf, token.NewFileSet(), expr.(*ast.StructType)); err != nil {
 			return
 		}
 
@@ -213,4 +202,14 @@ func optimalStructArrangement(sizes types.Sizes, m map[*types.Var]int) *types.St
 	})
 
 	return types.NewStruct(fields, nil)
+}
+
+func formatStruct(styp *types.Struct, curPkgPath string) string {
+	qualifier := func(p *types.Package) string {
+		if p.Path() == curPkgPath {
+			return ""
+		}
+		return p.Name()
+	}
+	return types.TypeString(styp, qualifier)
 }

--- a/testdata/src/struct/p.go
+++ b/testdata/src/struct/p.go
@@ -71,3 +71,11 @@ type s8 struct { // want `struct has size 40 \(size class 48\), could be 32 \(si
 	z uint32
 	w uint64
 }
+
+// Struct which combines multiple fields of the same type, see issue #41.
+type s9 struct { // want `struct has size 40 \(size class 48\), could be 24 \(size class 32\), you'll save 33.33% if you rearrange it to:\nstruct {\n\t_  \[0\]func\(\)\n\ti1 int\n\ti2 int\n\ta3 \[3\]bool\n\tb  bool\n}`
+	b      bool
+	i1, i2 int
+	a3     [3]bool
+	_      [0]func()
+}

